Enhancement proposal for Confidential Clusters

uril opened the pull request:
This enhancement proposes the integration of **confidential computing**
capabilities into **OpenShift cluster**, enabling the deployment of
**Confidential Clusters**

uril committed:
Enhancement proposal for Confidential Clusters

cgwalters commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,619 @@
+
+* RHEL CoreOS
+  * Add support for composefs (native), UKI, and systemd-boot to bootc (Bootable
+    Contrainers).
```
Containers

cgwalters commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,619 @@
+
+It then computes the expected PCR values for this bootable container image. It
+can either read a specific LABEL from the container image where those values
+have been pre-computed and stored, or pull the container image itself and
+directly compute the values.
```
I'd lean towards pull, but have a cache of course of container-sha ➡️ PCRs

uril replied:
If pre-calculated values are used, they must be validated (e.g. signed)

cgwalters reviewed (commented):
Awesome work overall! There's of course huge amounts of detail in some of this, but I think the outline looks good.

travier commented:
Direct link to rendered docs: https://github.com/uril/openshift_enhancements/blob/confidential-clusters-proposal/enhancements/security/confidential-clusters.md

travier commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,619 @@
+## Upgrade / Downgrade Strategy
+
+This component will (in the end) be part of the core OpenShift payload and
+update alongside the rest of the cluster.
```
```suggestion
updated alongside the rest of the cluster.
```

yuqi-zhang commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,619 @@
+    reference-values (expected "correct" values) in Trustee.
+
+* RHEL CoreOS
+  * Add support for composefs (native), UKI, and systemd-boot to bootc (Bootable
```
Since the MCO and on-cluster RHCOS operations doesn't currently use bootc at all, would that integration be needed here?

uril replied:
Details of MCO changes and the use of bootc-operator are postponed for later time

yuqi-zhang commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,619 @@
+    (cloud provider specific).
+  * Deploy the Confidential Cluster Operator on the bootstrap node
+
+* Confidential Cluster Operator
```
Will this be running as a core payload operator that's always present, or only deployed conditionally?

travier replied:
It will be an operator part of the core payload but only running if needed.

yuqi-zhang reviewed (commented):
Some general comments inline

travier commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,619 @@
+
+## Proposal
+
+Run all OpenShift nodes on Confidential VMs (CVMs). Use remote attestation to
```
Yes, the entire node runs as a Confidential VM provided by the cloud provider. You don't control on which host your VM runs (it's a cloud), and you can not run things outside.

travier commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,619 @@
+installer, passing in the URL of the external Trustee instance chosen above.
+1. The OpenShift installer generates a set of configuration files for the
+   external Trustee instance.
+1. If the cluster creator adds/removes/modifies MachineConfigs, the
```
The idea is that the config that will be used for the external Trustee server will include the full config passed to the bootstrap node. If anything modifies this config, then the Trustee config will have to be re-generated. We don't expect the manifests to be modified live during the installation.

uril replied:
To clarify, the point here is that the external Trustee config-files match the ignition-files generated by openshift-install. If the cluster-admin changes the cluster configuration, the external Trustee config-files need to change accordingly

uril committed:
FIXUP: add CVO capability

uril commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,619 @@
+This enhancement introduces some new API extensions:
+
+* **Running nodes on cloud CVMs**:
+For each supported cloud provider, confidential computing types and code need to
```
With regards to MAPI/CAPI - I think both should be able to request the cloud to run a CVM

uril committed:
FIXUP: the cocl-operator is a part of Core Payload

uril commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,620 @@
+includes a uniquely generated UUID. This UUID is given back to the new node. The
+operator watches for new Machine CRDs and sets up attestation and resource
+policy in the Trustee instance, and generates random secret values to be used as
+LUKS root keys.
```
It should work similarly for both MAPI and CAPI

uril committed:
FIXUP: fix some markdownlint complaints

- Fix title
- Add reviewers
- Fix some headers
- Add missing sections (e.g. DevPreview -> TechnPreview )

Still fails due to missing approvers

uril committed:
FIXUP: add introduction to confidential-computing and CVM

uril committed:
FIXUP: add Programming Languages - Rust and Go

uril commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,620 @@
+  * Setup a Trustee (attestation service) instance in the cluster to attest
+    nodes.
+  * Setup attestation and resource access policies in Trustee.
+  * Provide a registration server for new nodes to trigger the provisioning of
```
> Also, please look at CVO capabilities
I'll mention a CVO capability.

alicefr replied:
No, but they could be added eventually

alicefr commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,620 @@
+    attestation and fetch secrets upon a successful attestation.
+  * Add a clevis trustee pin to fetch LUKS passphrase upon a successful
+    attestation and encrypt/decrypt the disk.
+  * Modify Ignition to support clevis trustee pin.
```
we now allow arbitrary clevis pin in ignition ( see https://github.com/coreos/ignition/pull/2145) starting from ignition 3.6

uril committed:
Fixup: Add "Machine" and "ApprovedImage" CRDs

uril committed:
FIXUP: add a link to the bootstrap-phase design

dgoodwin commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,699 @@
+have been pre-computed and stored, or pull the container image itself and
+directly compute the values.
+
+The PCR pre-calculation flow is demonstrated in this presentation:
```
Can we get a definition of the PCR acronym in here somewhere.

UKI SEV-SNP could also use some clarification for those of us less familiar with the terminology.

dgoodwin commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,699 @@
+In a HCP scenario, the operator will only be responsible for the worker
+nodes. As the Confidential Cluster Operator will be hosted in the control plane,
+the nodes hosting those services are considered part of the Trusted Computing
+Base (TCB).
```
Could you spell out a theoretical real world scenario where this feature would make sense with a hypershift cluster. If my control plane is hosted by a provider, I wouldn't think there's much utility to having just the workers confidential. Is the assumption that the hypershift control plane cluster itself is a confidential cluster?

uril replied:
Yes, the assumption is that the cluster that is hosting the hypershift control plane is a confidential cluster.

I'll add this assumption to the document

dgoodwin commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,699 @@
+
+The protocol used by the nodes attestation client and the Trustee server must
+match. This means that we may have to keep multiple versions of the Trustee
+instance running in parallel until the boot image is updated in the cluster.
```
Who would orchestrate the multiple running versions of the trustee server? How complex will this be or do we have established precedent for similar?

alicefr commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,699 @@
+   Trustee PIN configuration.
+1. Cluster shutdown
+1. Before restarting any node, the Trustee instance must be made available at
+   the domain or IP configured above.
```
if trustee is down, then the attestation cannot be passed and the nodes fail to boot since they cannot retrieve the LUKS key of their root disk

alicefr commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,699 @@
+##### Making Confidential Cluster Operator fit OpenShift
+The following steps are to be taken to make cocl-operator fit OpenShift
+  * APIs and CRDs are to be written in Go
+  * When the operator is built, the APIs/CRDs are translated to Rust
```
No, it is pretty straightforward using Go -> yaml, then  yaml - kopium -> Rust

uril commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,619 @@
+In phase 2, the initial configuration will be modified to tell Ignition to fetch
+the new configuration from a remotely attested resource endpoint. The MCS will
+not serve Ignition configs directly for nodes anymore but will store those as
+resources in a Trustee instance. To access those configurations, the node will
```
The Trustee instance is being managed by the confidential-clusters-operator.
The operator is the one that would be asking MCS ignition configuration for new nodes and upload them as a secret to Trustee.
Alternatives are 1. The node gets encrypted ignition configs and request the key to decrypt them from Trustee. 2. The node requests a client-certificate and MCS only serves clients with a valid certificate.
I think in both cases, cocl-operator needs to talk with MCS and update Trustee, such that the information is available for the node only after a successful attestation.

JoelSpeed commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,620 @@
+* **ConfidentialCluster CRD**: This custom resource is used to configure the
+  Confidential Cluster Operator and indirectly the Trustee instance that is used
+  to attest nodes in the cluster and provide secrets.
+  It is namespaced, versioned and contains:
```
We have a lot of cluster scoped singleton resources for standlone clusters, this feels very much like it would be the same so I'd imagine rather than a future where you have multiple copies of the operator in multiple namespaces, it would be better to design the API first hand to consider a future with multiple trustees

What would the advantage be of deploying the operator multiple times?

uril committed:
FIXUP: add acronyms

uril committed:
FIXUP: HCP: the hosting cluster is trusted

uril committed:
FIXUP: add some info about Go to Rust conversion

uril commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,699 @@
+To make Confidential Cluster Operator a first citizen in the Openshift
+echosystem, interfaces are written in Go and generated with OpenShift tools.
+
+When the operator is built, the interfaces are converted to Rust. COPY FROM Jakob
```
Oops, I've updated this section.

uril commented:
/remove-lifecycle stale

uril committed:
Fixup: Add Confidential Cluster 'status' in infrastructure

uril committed:
Fixup: Workflow/Adding a new node: add reference to the implemetation

Add a reference to a subsection below with details about adding a new
node to the cluster.

uril committed:
Fixup: add link to related ignition PR

Configuring clevis-pin-trustee is now possible.

uril committed:
Fixup: mention MCO support for bootc based OS images

uril commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,699 @@
+
+##### Making Confidential Cluster Operator fit OpenShift
+The following steps are to be taken to make cocl-operator fit OpenShift
+  * APIs and CRDs are to be written in Go
```
No, only added them [here](https://github.com/trusted-execution-clusters/operator/tree/main/api)

uril committed:
Fixup: more info about Rust/Go and operator security

Also fix an unrelated typo

uril commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,699 @@
+The following steps are to be taken to make cocl-operator fit OpenShift
+  * APIs and CRDs are to be written in Go
+  * When the operator is built, the APIs/CRDs are translated to Rust
+  * The system openssl library is to be used
```
Building on Konflux and CVE scanning added.

uril commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,699 @@
+
+While we don’t want to support that for production, it should also be possible
+to test adding confidential nodes to a non confidential cluster where the
+Confidential Cluster Operator would be running, making testing easier.
```
E2E tests can be similar but with confidential clusters: installing a confidential cluster, scaling up/down worker nodes, rebooting some nodes, running a simple container, etc.

uril committed:
Fixup: add "security audit" before GA

uril commented:
/reopen

uril commented:
/reopen

uril commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,620 @@
+
+### Cluster installation
+
+As part of the cluster installation process in cloud platforms, a bootstrap node
+is created, which hosts a temporary control plane used to create the final
+control plane and worker nodes of the cluster.
```
Like all nodes, the bootstrap node gets its ignition config only after remote attestation.
The current proposal is to let Trustee serve the ignition configs to (all) nodes.
Alternatively it's possible to encrypt ignition-configs or use client-side certificates - either way the node must pass remote attestation.

uril commented on enhancements/security/confidential-clusters.md:
```diff
@@ -0,0 +1,619 @@
+node, which is considered trusted and it is used to bootstrap the trust for the
+rest of the cluster.
+
+In phase 2, the bootstrap node itself must be attested to establish trust. It is
```
Hosted Control Planes is mentioned in the document.
